tunnel/intra/doh/client_auth.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doh
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/eycorsican/go-tun2socks/common/log"
+)
+
+// CertificateLoader interface for requesting ClientAuth instances.
+type CertificateLoader interface {
+	// Request a ClientAuth instance (blocking).
+	// returns nil (no authentication) or a ClientAuth instance.
+	LoadClientCertificate() ClientAuth
+}
+
+// ClientAuth interface for providing TLS certificates and signatures.
+type ClientAuth interface {
+	// GetClientCertificate returns the client certificate (if any).
+	// It does not block or cause certificates to load.
+	// Returns a DER encoded X.509 client certificate.
+	GetClientCertificate() []byte
+	// GetIntermediateCertificate returns the chaining certificate (if any).
+	// It does not block or cause certificates to load.
+	// Returns a DER encoded X.509 certificate.
+	GetIntermediateCertificate() []byte
+	// Request a signature on a digest.
+	Sign(digest []byte) []byte
+}
+
+// clientAuthWrapper manages certificate loading and usage during TLS handshakes.
+// Implements crypto.Signer.
+type clientAuthWrapper struct {
+	loadCertificateOnce sync.Once
+	loader              CertificateLoader
+	signer              ClientAuth
+}
+
+func (ca *clientAuthWrapper) loadClientCertificate() {
+	// If no loader was provided then we can't load a certificate.
+	if ca.loader == nil {
+		log.Warnf("Client certificates are not supported")
```

Comment:
Some of these logs look like non-error conditions that should be logged at a lower severity.

Reply:
I think the only informative message is the one just below this.  The rest are unexpected.  I've reworded this one to make it more useful and made the next one Info.